### PR TITLE
fix(treesitter): improve vim.treesitter.foldexpr

### PR DESCRIPTION
**Pull request by tomtomjhj**
* Collect on_bytes and flush at the invocation of the scheduled callback
  to take account of commands that triggers multiple on_bytes.
* More accurately track movement of folds so that foldexpr returns
  reasonable values even when the scheduled computation is not run yet.
* Start computing folds from the line above (+ foldminlines) the changed
  lines to handle the folds that are removed due to the size limit.
* Shrink folds that end at the line at which another fold starts to
  assign proper level to that line.
* Use level '=' for lines that are not computed yet.

This fixes many issues I encountered with vim.treesitter.foldexpr, such as inconsistent foldlevels after editing and spurious opening of folds. It passes many interesting test cases I can think of, but probably needs a few more days of testing.

### CODE REVIEW

**Comment by lucario387** [comment]
Seems to fix the `<BS>` case for me :smile:

File: runtime/lua/vim/treesitter/_fold.lua [hunk 1]
```diff
@@ -5,35 +5,20 @@ local Range = require('vim.treesitter._range')
 local api = vim.api
 
 ---@class TS.FoldInfo
----@field levels table<integer,string>
----@field levels0 table<integer,integer>
----@field private start_counts table<integer,integer>
----@field private stop_counts table<integer,integer>
+---@field levels string[] the foldexpr value for each line
```
Comment (lewis6991) [inline runtime/lua/vim/treesitter/_fold.lua:8]: This is a bit ambiguous

```suggestion
---@field levels string[] the foldexpr result for each line
```

File: runtime/lua/vim/treesitter/_fold.lua [hunk 2]
```diff
@@ -361,6 +386,20 @@ function M.foldexpr(lnum)
   return foldinfos[bufnr].levels[lnum] or '0'
 end
 
+function M.refresh()
```
Comment (lewis6991) [inline runtime/lua/vim/treesitter/_fold.lua:389]: Does this need to be exposed? If it does can be prefix it with `_` so it doesn't need to be maintained under any API contract?

**Review by lewis6991 — APPROVED** [review]
Haven't studied this closely, but the code looks well written with appropriate comments in places and the test coverage is significantly increased.

**Comment by justinmk** [comment]
Some folds tests are failing in CI

File: runtime/lua/vim/treesitter/_fold.lua (hunk 2 above)
Comment (tomtomjhj) [reply runtime/lua/vim/treesitter/_fold.lua]: Not really. Inlined the function to the OptionSet autocmd.

**Comment by tomtomjhj** [comment]
> subtle LuaJIT difference in either format

fixed. https://github.com/neovim/neovim/compare/d2e6a4e9325b5c003b4c49d3cbc0a7d4f602da5e..6f75facb9d755e2e3a10a9a1d6d258a12578dc14#diff-8ac3a3b95aa90eda1236a1d02af0034db91bf22442c4b77272c94e9873439b89R43

**Comment by igorlfs** [comment]
Hello!

I was tracking down a bug I noticed in a [plugin](https://github.com/NeogitOrg/neogit/issues/1521) and I believe it's actually an issue with core, and this PR is related.

## Problem

The problem is that setting the `'foldminlines'` option **for a given window** actually updates the folds for all windows **using the newly set** `'foldminlines'`, **in spite of it not having changed for the other windows**.

I can create a new issue if here isn't the appropriate place to have this discussion, but in the meantime, here's a repro:

## Minimal Config

```lua
vim.api.nvim_create_autocmd("FileType", {
	callback = function(args)
		vim.treesitter.start(args.buf)

		vim.api.nvim_buf_call(args.buf, function()
			vim.wo.foldexpr = "v:lua.vim.treesitter.foldexpr()"
			vim.wo.foldmethod = "expr"
			vim.wo.foldcolumn = "1" -- makes it easier to notice that the folds were incorrectly updated
		end)
	end,
})
```

## Files

`foo.lua`

```lua
return {}
```

`bar.lua`

```lua
return {
	fizz = {},
	buzz = {}
}
```

## Repro (nightly)

1. `nvim -u minimal.lua foo.lua`
2. `<C-w>v`
3. `:e bar.lua`
4. `zO`
5. `<C-w>l`
6. `:lua vim.wo[0].foldminlines=0` -- updates **only** `foo.lua`'s `'foldminlines'`

And then, `bar.lua`'s folds will also be updated with `'foldminlines' = 0`

**Comment by lewis6991** [comment]
Please do not raise issue reports on the end of merged PR's

**Pull request merged**